Full decoupling of layout algorithm and rendering algorithm

Perry26 opened the pull request:
Makes the drawing algorithm fully independent of the force simulation and allows for arbitrary lay-out algorithms, including but not limited to the ones generated by d3-forces.

christojeffrey commented:
nice! I noticed that all the simulation is gone right? now we kinda use d3 just for svg manipulator. I guess your goal (based on this PR title and description and the changes) is to take out all the way d3 manipulate the location (x and y) of the data, and just create all those manipulator ourselves. Please let me know if I understand your goal correctly. 

I've checked the preview and notice that some of our previous features is not working. For example, the children node no longer determine the size of the parent node. I assume you're not going to fix all of it in this PR, so do you mind listing all of it? Please also list any features that you see as not feasible to be done in this new structure if you notice them. I'll try to implement them to make this new structure as featureful as the previous one. Please let me know also whether you agree or not with that plan (for example maybe you prefer to drop some of the previous features)

Perry26 commented:
The thing of the child node not determining the size of the parentnode is actually a feature and not a bug. It is, to be percise, the size of the child _layout_ that determines the size of the parentnode. 
Currently you cannot change the size of the child layout after generation, and I'm not entirely sure we'd want to change that, because it would make the system behave rather unpredictibly. Since some layouts will be statically computed, the cannot nececarlily be updated gradually like force based layouts (or at least not with a lot extra work, you'd need to use special techniques in the layout implementation).

Anyhow, here are all design changes listed in no particular order (hoping I didn't forget anything):
 - Add general helper function notNaN to prevent NaN's and infinities from "sneaking into" the code. (You'd be surprised how much of an issue this was when making a custom force; function is to be called on any value that gets passed to the layout data.)
 - All nodes use center coordinates instead of top-left coordinates.
 - Draw function no longer depends on config
 - Drag and Drop now works based on mouse movement instead of target coordinate, and will be "restricted" to the current layout size. Also no longer does any collision detection / no longer moves other nodes out of the way.
 - I perserved the old gradient code, and most of the edge lifting / filtering code
 - The draw function provides a callback that can be used to update the drawing without rerendering the layout. Some drawsettings updates are hence only rerendering stuff, and some are "relayouting" stuff
 - Layouting is completely separated from rendering, and not dependent on d3-forces
     - The function splitnodes splits the nodes into 4 different arrays, (rootnodes, innernodes, intermediatenodes, and simplenodes). Every array of nodes can be layouted independently, provided the inner level is strictly layouted before the outer level.
     - After calculating the layout on all arrays, the rendering functions are called.
     - Layouting is done in 2 phases:
         - Every simple node (node with no child nodes) gets a set width and height (drawSettings.minimumNodeSize)
         - For every parentnode, all **child** nodes get assigned an x and y coordinate from the layouting algorithm, and the (one) **parent** node get a widht and height form the algorithm
     - The following layouts are implemented: Circular layout (innernodes only), straightTree, forceBasedLayout
     - All layouts are exported by the file "Layouts"
 - Once again, I changes the convert and filter to keep extra information / a data structure
     - The converter now gives a nodesDictonary to the drawer. I already implemented this once wayback, and I'm not sure when it was removed, but It's back because I kindoff needed it to do stuff.
     - The edgelifter now annotates each edge with information about where it would be lifted if everything was globally lifted at level 0, because this is the only way you can get a source/target that is guaranteed to be in the actual relevant layout. These are also pretty much the only properties used by the layouting algorithm
 - GraphDataEdge now explicitly either has sources and targets by id _or_ by reference, since the d3 force package might change this.
 - No interfaces inherit from d3-forces currently.



Some further thoughts on things we might want to consider, and some downsides to this architecture:

 - I switched to center coordinates instead of top-left coordinates for nodes, because I was reworking everything anyway and used a force-based layout as the default. However, for some other algorithms top-left seems more convenient (specifically the straight tree algorithm). To solve that ...
 - ... we might want to use classes for some datastructures rather than interfaces, specifically the GraphDataNode. 
     - For example, we could pass everything through notNaN automatically.
     - ... and this could settle the topleft vs center coordinate problem. Just pick one for the implementation of the class internally, but offer a facade for the other coordinate system.
 - Do we want to fix drag and drop / reimplement collision detection and node resizing at all? There are techniques other than using force-simulations to make it possible to update a layout dynamically, but it's an entirely different kettle of fish.
 - Drawsettings is overloaded and does 2 unrelated things: We might want explicitly split the "relayouting" from the rerendering in the architecture. 
 - The rendering algorithm presumes all layouts to be centered at point 0,0, but this is not nesecarily the case (most obviously in the force-based layout). We might want to introduce some intermediate stage to center the layout before passing it to the parent.
 - Splitnodes is a bit wonky, because I'm not sure what to do with simpleNodes at the top level. It feels like this mechanism is a bit ill-designed by me but I'm not sure how to improve it.
 - Obviously, we want a ui for picking the layout (currently these are hard-coded in the draw function)

christojeffrey commented:
well that's a lot :D it's good to see some reasoning behind your decision. about the nodesDictionary ... I removed it. Because we have list of nodes already (I forgot the name of the attribute for this) and they have exactly the same data (just with different structure) "why not remove the list of node instead?" becuse d3 accept array for the simulation, not an object. part of me wants to stop doing another major refactor ahaha, probably you feel this too. But I understand it's probably needed. I'll try to look at drawSettings after this, based on your comment above. rethinking the structure of drawSettings probably will bleed into the main structure though. I hope it's fine, let me know if you have any concern about this. 

I don't have any suggestion to make regarding draw function, I guess removing d3 forces and simulation is a good move here. hopefully it won't bite us later. approved! good job 👍
